mne/preprocessing/xdawn.py

```diff
@@ -0,0 +1,412 @@
+
+    window = nmax - nmin
+    ne, ns = data.shape
+    to = {}
+
+    for eid in event_id:
+        # select events by type
+        ix_ev = events[:, -1] == event_id[eid]
+
+        # build toeplitz matrix
+        trig = np.zeros((ns, 1))
+        ix_trig = (events[ix_ev, 0] / decim) + nmin
+        trig[ix_trig] = 1
+        toep = linalg.toeplitz(trig[0:window], trig)
+        to[eid] = toep
+
+    # Concatenate toeplitz
+    to_tot = np.concatenate(to.values())
+
+    # least square estimation
+    evo = np.dot(np.dot(linalg.pinv(np.dot(to_tot, to_tot.T)), to_tot), data.T)
+
+    # parse evoked response
+    evoked_data = {}
+    for i, eid in enumerate(event_id):
+        evoked_data[eid] = np.array(evo[(i * window):(i + 1) * window, :]).T
+
+    return evoked_data, to
+
+
+def _check_overlapp(epochs):
+    """check if events are overllaped."""
+    isi = np.diff(epochs.events[:, 0])
+    window = int((epochs.tmax - epochs.tmin) * epochs.info['sfreq'])
+    return isi.min() < window
+
+
+def _construct_signal_from_epochs(epochs):
+    """Reconstruct pseudo continuous signal from epochs."""
+    start_ix = (np.min(epochs.events[:, 0])
+                + int(epochs.tmin * epochs.info['sfreq']))
+    end_ix = (np.max(epochs.events[:, 0])
+              + int(epochs.tmax * epochs.info['sfreq']) + 1)
+
+    ns = end_ix - start_ix
+    ns_epochs = epochs._data.shape[2]
+    ne = epochs._data.shape[1]
+    ix_events = epochs.events[:, 0] - epochs.events[0, 0]
+
+    data = np.zeros((ne, ns))
+    for i, ix in enumerate(ix_events):
+        data[:, ix:(ix + ns_epochs)] = epochs._data[i]
+
+    return data
+
+
+def least_square_evoked(epochs, return_toeplitz=False):
+    """Least square estimation of evoked response from a epoch instance.
+
+    Parameters
+    ----------
+    epochs : Epoch object
+        An instance of Epoch.
+    return_toeplitz : bool
+        if true, compute the toeplitz matrix
+
+    Returns
+    -------
+    evokeds : dict of evoked instance
+        An dict of evoked instance for each event type in epochs.event_id.
+    """
+    if not isinstance(epochs, _BaseEpochs):
+        raise ValueError('epochs must be an instance of `mne.Epochs`')
+
+    evs = epochs.events.copy()
+    evs[:, 0] -= evs[0, 0] + int(epochs.tmin * epochs.info['sfreq'])
+    data = _construct_signal_from_epochs(epochs)
+    evo, to = _least_square_evoked(data, evs, epochs.event_id,
+                                   tmin=epochs.tmin, tmax=epochs.tmax,
+                                   sfreq=epochs.info['sfreq'], decim=1)
+    evokeds = {}
+    info = cp.deepcopy(epochs.info)
+    for name, data in evo.items():
+        n_events = len(evs[evs[:, 2] == epochs.event_id[name]])
+        evoked = EvokedArray(data, info, tmin=epochs.tmin,
+                             comment=name, nave=n_events)
+        evokeds[name] = evoked
+
+    if return_toeplitz:
+        return evokeds, to
+
+    return evokeds
+
+
+class Xdawn():
+
```

Comment:
According to D203, 1 blank line required before class docstring. 
I can remove it but flake8 will not be happy